src/components/Table/Table.tsx

```diff
@@ -0,0 +1,186 @@
+import type {FlashListRef} from '@shopify/flash-list';
+import React, {useImperativeHandle, useRef, useState} from 'react';
+import TableContext from './TableContext';
+import type {TableContextValue, UpdateFilterCallback, UpdateSortingCallback} from './TableContext';
+import type {ActiveSorting, GetActiveFiltersCallback, GetActiveSearchStringCallback, GetActiveSortingCallback, TableHandle, TableMethods, TableProps, ToggleSortingCallback} from './types';
+
+// We want to allow the user to switch once between ascending and descending order.
+// After that, sorting for a specific column will be reset.
+const MAX_SORT_TOGGLE_COUNT = 1;
+
+function Table<T, ColumnKey extends string = string, FilterKey extends string = string>({
+    ref,
+    data = [],
+    columns,
+    filters,
+    compareItems,
+    isItemInFilter,
+    isItemInSearch,
+    children,
+    ...listProps
+}: TableProps<T, ColumnKey, FilterKey>) {
+    if (!columns || columns.length === 0) {
+        throw new Error('Table columns must be provided');
+    }
+
+    const [currentFilters, setCurrentFilters] = useState<Record<string, unknown>>(() => {
```

Comment:
It feels like there is a lot of logic encapsulated into this one component. Thinking of the future, if someone were to add more functionality to the table, then this component would be growing "vertically" and not "horizontally".

cc @adhorodyski

Reply:
I've extracted filtering/sorting/searching into "middleware hooks", that each follow a `Data[] -> middleware() -> Data[]` scheme